[Compose 페이먼츠 2단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
```
2단계 리뷰도 잘 부탁드립니다 🙇‍♂️
```

```
지오, Modifier.testTag에 대해 어떻게 생각하시나요?
이번에 테스트 코드를 작성하면서 contentDescription과 Role을 활용해보려고 했는데, 컴포넌트 접근이나 확인이 생각보다 쉽지 않더라고요.

생각해보면 contentDescription도 결국 TalkBack 같은 접근성 기능을 위해 로컬라이징 처리가 필요합니다.
그렇다면 테스트 코드에서 컴포넌트를 식별하기 위해 문자열할 때의 문제와 동일한 문제가 발생하는 것 같습니다.
그래서 저는 testTag()를 사용하는 것이 테스트를 위해 필요한 리소스를 줄일 수 있다고 생각합니다.

많은 글에서 프로덕션 코드에 테스트 용 코드가 포함되면 코드가 오염된다고 표현하고 있는데,
testTag() 한 줄 정도가 프로덕션 코드에 오염이라고 할 만큼 큰 영향을 줄 수 있는지는 아직 잘 모르겠습니다.
```

https://proandroiddev.com/stop-using-test-tags-in-the-jetpack-compose-production-code-b98e2679221f
https://callmeryan.medium.com/efficient-testing-or-code-pollution-the-role-of-testtag-in-jetpack-compose-testing-4892ee73e966

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github.com/user-attachments/assets/81030f39-4dbe-4fc4-8713-448d18f02d7e

moondev03 committed:
refactor: MainActivity를 CardRegistrationActivity로 변경

moondev03 committed:
test: 만료일 입력창 테스트 케이스 수정

- 만료일 월 입력값 검증 테스트에서 예외 발생 대신 에러 메시지 노출을 확인하도록 수정
- 만료일 연도 입력값 검증 테스트에서 유효하지 않은 만료일 입력 시 에러 메시지 노출을 확인하도록 수정

moondev03 committed:
test: 카드 번호 입력창 테스트 코드 개선

- AssertAll을 사용하여 테스트 케이스를 묶어 관리하도록 변경

moondev03 committed:
test: GroupedVisualTransformationTest 클래스명 변경

moondev03 committed:
refactor: 각 입력창 모델 구현

- CardNumberUiModel
- CardExpirationDateUiModel
- CardPasswordUiModel
- CardholderNameUiModel
- 각 입력창 모델의 유효성 검증 로직 구현

- 각 입력창 Composable 함수 수정
- 입력값을 각 모델로 받도록 수정
- 모델의 유효성 검증 결과를 바탕으로 에러 메시지 표시 로직 추가

moondev03 committed:
refactor: CardRegistrationScreenUiState 필드 변경

moondev03 committed:
test: 테스트 코드 수정

moondev03 committed:
docs: 2단계 기능 요구 사항 작성

moondev03 committed:
refactor: 카드 유효성 검증 로직 제거

- 카드 유효성 검증 관련 인터페이스 및 구현체를 제거합니다.

moondev03 committed:
feat: 카드 목록 화면 상단 앱 바 구현

moondev03 committed:
feat: 카드 정보 표시 기능 구현

moondev03 committed:
feat: 카드 등록 버튼 구현

moondev03 committed:
feat: getParcelableExtraCompat 구현

moondev03 committed:
feat: 카드 등록 기능 구현

- 카드 등록 화면 구성
- 카드 등록 성공 시, 등록된 카드 정보를 이전 화면으로 전달

moondev03 committed:
feat: 카드 목록 화면 구현

- 카드를 추가하고 목록을 볼 수 있는 화면을 구현했습니다.
- 카드가 없을 때, 하나 있을 때, 여러 개 있을 때 각각 다른 UI를 보여줍니다.
- 카드 추가 화면으로 이동하는 기능을 구현했습니다.
- 카드 추가 후 토스트 메시지를 보여줍니다.

moondev03 committed:
refactor: CardsScreen 코드 구조 개선 및 UI 메시지 문자열 리소스화

- CardsScreen 내 중복 코드 제거 및 로직 간결화
- 카드 등록 결과 메시지 및 카드 없을 시 안내 메시지를 strings.xml로 분리
- CardsScreenPreviewParameterProvider에 DUMMY_CARDS 상수 추가하여 가독성 향상

moondev03 committed:
refactor: StringResWithParams 삭제

- StringResWithParams 클래스를 삭제하고, 안드로이드 기본 리소스 처리 방식을 사용하도록 변경합니다.

moondev03 committed:
test: PaymentCard 컴포넌트 테스트 구현

moondev03 committed:
test: CardsScreen, CardRegistrationScreen 테스트 코드 추가

moondev03 committed:
docs: 카드 목록 화면 구현 요구사항 완료 표시

moondev03 committed:
refactor: 이벤트 소비 후 null로 초기화

- uiEvent를 소비한 후 null로 초기화하여 중복 소비를 방지

moondev03 committed:
test: 빈 목록 상태 테스트에서 안내 문구가 실제로 검증되지 않는 문제 수정

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardRegistrationButton.kt:
```diff
@@ -0,0 +1,51 @@
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    Box(
+        modifier =
+            modifier
+                .shadow(8.dp, RoundedCornerShape(5.dp))
+                .size(width = 208.dp, height = 124.dp)
+                .background(
+                    color = Color.LightGray,
+                    shape = RoundedCornerShape(5.dp),
+                ).clickable(onClick = onClick),
+        contentAlignment = Alignment.Center,
+    ) {
+        Icon(
+            imageVector = Icons.Default.Add,
+            contentDescription = stringResource(R.string.cards_screen_registration_button_content_description),
+            tint = Color.Black,
+        )
```
왜 부모 노드에 contentDescription을 위임해야 하나요?

moondev03 committed:
test: ParameterizedTest의 이름 포맷 변경

- ParameterizedTest의 이름을 `groupSize={0}, separator='{1}', input='{2}' => '{3}'`에서 `[{index}] {arguments}`로 변경하여 가독성을 개선했습니다.

moondev03 committed:
refactor: 카드 소유지 이름이 긴 경우 처리

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -75,32 +60,23 @@ fun CardExpirationDateTextField(
+        cardExpirationDate = cardExpirationDate,
+        onCardExpirationDateChanged = {},
+    )
+}
```
Preview는 잘 보이지만, Interactive Mode 또는 Run Preview를 통해 입력이 잘 되는지는 확인하기 어려운 것 같아요. 😢 

어떻게 개선해볼 수 있을까요?

moondev03 replied:
`Interactive Mode`나 `Run Preview`는 유용한 기능이지만, 단순히 입력에 따른 상태 변화를 확인하는 목적이라면 꼭 필요하지는 않다고 생각합니다. 현재도 PreviewParameter를 통해 상태별 컴포넌트의 모습을 충분히 확인할 수 있다고 생각해요.

입력에 따른 상태 변화 검증은 오히려 UI 테스트에서 다루는 것이 더 적절하다고 생각합니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateUiModel.kt:
```diff
@@ -0,0 +1,45 @@
+
+    @IgnoredOnParcel
+    private val yearMonth: YearMonth? =
+        runCatching { YearMonth.parse(cardExpirationDate, formatter) }.getOrNull()
```
`yearMonth`가 nullable인 이유는 파싱에 실패할 수 있기 때문이라고 보여요. 하지만 이 또한 `init` 블록에서 검증할 수 있을 것 같습니다. 👀

moondev03 replied:
검증 로직을 도메인 모델로 분리하고, Factory 패턴을 활용하여 해당 검증이 수행되도록 수정하였습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateUiModel.kt:
```diff
@@ -0,0 +1,45 @@
+@Parcelize
+data class CardExpirationDateUiModel(
+    val cardExpirationDate: String,
+) : Parcelable {
```
이 클래스는 `cardExpirationDateUiModel` 또는 `cardExpirationDate` 형식으로 인스턴스화 될 것 같아요.

갖고 있는 값에 접근할 때 `cardExpirationDate.cardExpirationDate` 형식으로 하면 중복되니, `value`라는 이름도 고려해볼만 한 것 같아요. 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -23,17 +35,81 @@ fun PaymentCard(modifier: Modifier = Modifier) {
+                    shape = RoundedCornerShape(4.dp),
                 ),
+    )
+}
```
개인적으로는 이렇게 분리된 컴포저블도 프리뷰로 확인할 수 있으면 좋을 것 같아요. 👀

moondev03 replied:
추가했습니다!

[53a70b5](https://github.com/woowacourse/android-payments/pull/34/commits/53a70b5527df3f4d1121d19e3abce353509223e9)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -23,17 +35,81 @@ fun PaymentCard(modifier: Modifier = Modifier) {
+private fun PaymentCardInfo(
+    paymentCardUiModel: PaymentCardUiModel?,
+    modifier: Modifier = Modifier,
+) {
```
인자 `paymentCardUiModel`의 자료형이 nullable하다는 점이 혼란스러워요. `paymentCardUiModel`이 없는 경우에 `PaymentCardInfo`를 표시할 수 있을까요?

반대로 생각해서 **있는 경우에만 보여주는 방식**을 선택해볼 수 있을 것 같습니다. 🙂

moondev03 replied:
있는 경우에만 보여주도록 하는게 더 적절한 것 같네요 :)

[53a70b5](https://github.com/woowacourse/android-payments/pull/34/commits/53a70b5527df3f4d1121d19e3abce353509223e9)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -23,17 +35,81 @@ fun PaymentCard(modifier: Modifier = Modifier) {
+            expirationDate = CardExpirationDateUiModel("1224"),
+            cardholderName = CardholderNameUiModel("JOHN DOE"),
+        )
+    PaymentCard(paymentCardUiModel = paymentCardUiModel)
```
`paymentCardUiModel`이 `null`인 경우도 궁금해요! 👀

moondev03 replied:
카드 등록 화면에서는 PaymentCardUiModel를 생성하지 않았을 때도 카드 자체는 보여주기 위해서 null을 받도록 했었던 것 같습니다. nullable 하지 않아도 처리될 수 있도록 수정했습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCardUiModel.kt:
```diff
@@ -0,0 +1,32 @@
+                    0, 1 -> chunk
+                    else -> mask
+                }
+            }.joinToString(DEFAULT_NUMBER_SEPARATOR)
```
`mask` 인자를 열어두신 이유가 무엇일까요? 👀

moondev03 replied:
화면마다 카드를 보여줄 때 마스킹되는 범위나 문자가 달라질 수 있기 때문에
인자로 받아 사용할 수 있도록 구현했습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCardUiModel.kt:
```diff
@@ -0,0 +1,32 @@
+    fun formattedExpirationDate(separator: String = DEFAULT_EXPIRATION_DATE_SEPARATOR): String =
+        expirationDate.cardExpirationDate.chunked(2).joinToString(separator)
+
+    val upperCardholderName: String = cardholderName.cardholderName.uppercase()
```
`uppercase`로 반환하는 것은 어디에서 책임지면 좋을까요? 🤔 

지금은 두 군데(`PaymentCardUiModel`, `CardholderNameTextField`)에서 해당 작업을 해주고 있는 것 같아요. 👀

moondev03 replied:
`PaymentCardUiModel`에서 책임져야 할 것 같습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,165 @@
+            }
+
+            else -> Unit
+        }
```
`CardsScreenUiEvent`를 `sealed interface`로 만들어주셨는데, `else`를 쓰면 그 장점을 잃을 수 있을 것 같습니다. 😢 

`else`를 쓰지 않고 작성해볼 수 있을까요? 🤔

moondev03 replied:
👍 👍 👍 

[dbd8189](https://github.com/woowacourse/android-payments/pull/34/commits/dbd8189b352dbe96f5357f5987f9063474deba2d)

giovannijunseokim replied:
제가 의도한건 `else -> ...` 대신 `null -> ...`을 활용하는 방법이었는데, 다이스의 개선된 코드도 괜찮네요. 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,165 @@
+            is CardsUiState.SINGLE ->
+                CardsSingleContent(cardsUiState.card, onRegistrationButtonClick)
+        }
+    }
```
만약 `CardsMultipleContent`에서 `Column` 대신 다른 레이아웃 컴포저블을 사용하도록 변경사항이 생긴다면 어떻게 대처할 수 있을까요? 🤔 

`Column`을 사용하는 것이 `CardsScreenContent`의 책임일까요?

moondev03 replied:
지오의 말을 듣고 개선해봤습니다 :)

[3944695](https://github.com/woowacourse/android-payments/pull/34/commits/3944695a847f902a2f401bc9f8ba100ccd752c4d)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/cards/CardsTopAppBar.kt:
```diff
@@ -0,0 +1,67 @@
+        actions = { if (isRegistrationButtonEnabled) RegistrationButton(onRegistrationButtonClick) },
+        modifier = modifier,
+    )
+}
```
인자 이름은 `~~~Enabled` 형식이지만, 실제로 그 목적으로 쓰이고 있지 않은 것 같아요. 👀

moondev03 replied:
수정했습니다 !

[2ca9632](https://github.com/woowacourse/android-payments/pull/34/commits/2ca9632697f0cbc5fec29d74b54f2076f1b2eb92)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/cards/CardsTopAppBar.kt:
```diff
@@ -0,0 +1,67 @@
+    onRegistrationButtonClick: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    TextButton(
+        onClick = onRegistrationButtonClick,
+        colors =
+            ButtonDefaults.textButtonColors(
+                contentColor = Color.Black,
+                disabledContentColor = Color.Gray,
+            ),
+        modifier = modifier,
+    ) {
+        Text(
+            text = stringResource(R.string.cards_screen_top_app_bar_registration_button_text),
+            fontWeight = FontWeight.Bold,
+        )
+    }
```
`Text` 컴포저블에서 `Modifier.clickable`을 사용하는 것과 어떤 차이가 있나요? 👀

moondev03 replied:
찾아봤는데 아래와 같이 구분해서 사용하는 것 같아요!

```
버튼 같은 “액션” → TextButton
문장 중 특정 단어를 클릭할 수 있게 만들고 싶을 때 → Text + clickable
```

추가로 Button으로서 제공하는 기능들을 활용할 수 있기 때문에 TextButton을 사용했습니다.
ex. 터치 영역 보장, 시멘틱 의미, enabled 등등

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreen.kt:
```diff
@@ -8,49 +8,46 @@ import androidx.compose.foundation.layout.padding
+fun CardRegistrationScreen(
+    viewModel: CardRegistrationScreenViewModel = rememberCardRegistrationScreenViewModel(),
+    onBackClick: () -> Unit,
+    onRegistrationComplete: (PaymentCardUiModel) -> Unit,
```
의도된 인자 순서일까요? 👀

moondev03 replied:
의도하지는 않았지만 지금보니 뒤로가기가 아래로 내려가는게 자연스러울 것 같네요 :)

giovannijunseokim replied:
기본값이 존재하는 인자와 그렇지 않은 인자간의 순서에 대한 이야기였습니다! 😅

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenViewModel.kt:
```diff
@@ -3,60 +3,46 @@ package woowacourse.payments.ui.screen.registration
-        val message =
-            StringResWithParams(R.string.card_registration_screen_registration_card_success)
-        uiEvent = CardRegistrationScreenUiEvent.ShowSnackbar(message)
+        if (!_uiState.isSaveButtonEnabled) return
```
이렇게 단순히 값을 얻을 때는 `_uiState` 대신 `uiState`를 사용할 수도 있을 것 같은데요. 👀 

다이스는 어떤 어떤 기준으로 `uiState`와 `_uiState`를 나눠서 사용하시나요?

moondev03 replied:
저는 `외부/내부`로 구분하여 사용하고 있어요!
`읽기/쓰기` 처럼 구분할 수 있을 것 같긴한데 `get()`을 거쳐서 접근할 이유를 느끼지 못했던 것 같아요.

공식 [샘플 코드](https://github.com/android/compose-samples/blob/main/Reply/app/src/main/java/com/example/reply/ui/ReplyHomeViewModel.kt)에서는 이렇게 사용하네요 🤔
```kotlin
val email = uiState.value.emails.find { it.id == emailId }
_uiState.value = _uiState.value.copy(
    openedEmail = email,
    isDetailOnlyOpen = contentType == ReplyContentType.SINGLE_PANE,
)
```
```kotlin
val currentSelection = uiState.value.selectedEmails
_uiState.value = _uiState.value.copy(
    selectedEmails = if (currentSelection.contains(emailId)) currentSelection.minus(emailId) else currentSelection.plus(emailId),
)
```

지오는 어떻게 구분해서 사용하고 있나요?

giovannijunseokim replied:
저는 읽기/쓰기로 구분하는 편입니다. _uiState와 uiState의 차이는 불변/가변인데, 값을 얻을 때엔 불변에게서 얻어오는 것이 맘이 편하더라구요.

다만 내부적으로는 거의 동일한 동작이라 큰 상관은 없을 것 같아요. 😁

giovannijunseokim reviewed (changes requested):
안녕하세요 다이스. 미션 구현 고생하셨습니다! 👍 

궁금한 점들을 리뷰 내용으로 남겨두었으니 확인 후 답변 부탁드립니다. 😁 

궁금한 점이나 논의하고 싶은 내용이 있으면 편하게 얘기해주세요! 🙏

> 지오, Modifier.testTag에 대해 어떻게 생각하시나요?

`testTag()`가 프로덕션 코드에 오염을 일으키는가 이전에,
**`테스트코드를 위한 코드가 프로덕션 코드를 오염시키는가?`** 에 대해 생각해보아야 할 것 같아요.

제 의견은 **`오염이라 볼 수 없다`** 입니다. 🙂

1. `Repository`에서 파생되는 `DefaultRepository`, `FakeRepository`는 결국 테스트를 위한 추상화 아닐까요? 이 또한 오염이라 할 수 있을까요?
2. `LocalDateTime.now()` 등의 의존하는 도메인 모델의 경우 optional parameter를 활용하여 테스트를 용이하도록 합니다. 이 또한 오염일까요?

그 외에도 여러 예시가 있겠지만, 이 정도만 생각해보아도 이미 프로덕션 코드는 테스트를 위한 코드가 많이 존재한다고 생각해요. 🤔 

정답이 있는 문제라기보단, **테스트를 얼마나 중요하게 여기는지**에 따라 답이 달라질 것 같네요. 😁

woowahan-leah commented:
@giovannijunseokim 
다이스에게 해줄 수 있는 **설계** 측면에서의 피드백도 있을까요?

giovannijunseokim commented:
@woowahan-leah 
레아가 말씀하시는 설계 측면이란 패키지 설계 및 `UiState`, `UiModel` 설계를 말씀하시는 걸까요? 👀 

Jetpack Compose를 사용할 때의 프로젝트 설계에 대해서는 저도 모르는 부분이 많아서, 우선 **강의에서 다룬 내용을 중점**으로 코멘트를 남겨보았습니다. 😁

레아의 코멘트를 보고 다시 한 번 프로젝트를 살펴보니 **`UiModel` 객체에 의해 코드 복잡성이 높아진 느낌**이 있네요.

@moondev03 
```
값에 대한 검증을 처리하기 위해 `XXXUiModel` 형태의 객체가 늘어났는데,
이는 `UiModel`의 책임일까요? 🤔 

또한 거의 컴포저블 하나당 하나의 `UiModel`을 사용되었기에
`Composable` 코드의 복잡성도 늘어난 듯한 느낌도 듭니다. 😢 

사용하는 자료형이 `String`에서 `XXXUiModel`로 변하면서,
해당 컴포저블을 처음 보는 사람은 `UiModel`까지 확인해야 한다는 한 층의 뎁스가 생긴 것 같아요.
```

추가적으로 레아께서 제가 남긴 코멘트 중 **특정 부분에 대한 코멘트가 부족하다**고 생각되신다면,
어떤 부분인지 알려주실 수 있나요? 🙇

woowahan-leah commented:
@giovannijunseokim 
좋은 의견이네요~! @moondev03 의 의견도 듣고 싶네요.
그리고 혹시 ViewModel 구조에 대해서는 어떻게 생각하시나요?
이 구조가 별로니 바꿔라! 라는 이야기를 하고 싶은게 아니라, 리뷰어로서의 의견을 여쭙고 싶은 것이니 부담 갖지 마시고요 ☺️

giovannijunseokim commented:
리뷰어로서 다이스가 작성해준 ViewModel들에 대해 이야기 나눠보고 싶은 점들은 다음과 같습니다. 

1. `ViewModel`은 `Activity`에서 들고 있어야 하는가, Screen(`Composable`)에서 들고 있어야 하는가?
2. `ViewModel`에서 `androidx.compose.runtime`에 대한 의존성을 갖고 있어도 될까?

포인트들을 적고 난 후 제 생각을 정리해 보았습니다. 🤔 

다만 이는 다이스가 먼저 공유해주신 후 제 의견을 공유하는 것이 더 좋을 것 같다는 생각이 듭니다. 😁
코멘트로 달아주시면 감사드리겠습니다! 🫡 @moondev03 

레아 덕분에 다시 한 번 생각해볼 수 있었습니다. 감사합니다! 🙇 @woowahan-leah

moondev03 committed:
refactor: UI 모델 패키지 위치 변경

- `component` 패키지에서 `model` 패키지로 UI 모델 클래스 이동
- 관련 import 경로 수정

moondev03 committed:
refactor: 문자열 리스소 정리

moondev03 committed:
refactor: 카드 만료일 유효성 검증 로직 변경

- `CardExpirationDate` 도메인 클래스에서 유효성 검증 로직을 담당하도록 변경
- `CardExpirationDateUiModel`은 UI 상태만 관리

moondev03 committed:
refactor: 카드 소유자 이름 유효성 검증 로직 변경

- `CardholderName` 도메인 클래스에서 유효성 검증 로직을 담당하도록 변경
- `CardholderNameUiModel`은 UI 상태만 관리

moondev03 committed:
refactor: 카드 번호 유효성 검증 로직 변경

- `CardNumber` 도메인 클래스에서 유효성 검증 로직을 담당하도록 변경
- `CardNumberUiModel`은 UI 상태만 관리

moondev03 committed:
refactor: 카드 비밀번호 유효성 검증 로직 변경

- `CardPassword` 도메인 클래스에서 유효성 검증 로직을 담당하도록 변경
- `CardPasswordUiModel`은 UI 상태만 관리

moondev03 committed:
fix: Preview 누락된 정보 추가

moondev03 committed:
refactor: 유효성 검사 로직 ViewModel로 이동

moondev03 committed:
feat: 카드 비밀번호 입력창에 `visualTransformation` 매개변수 추가

moondev03 committed:
style: CardholderNameTextField, CardNumberTextField, PaymentCard 에서 불필요한 코드 제거

moondev03 committed:
test: 유효한 카드 정보 fixture 추가

moondev03 committed:
test: 카드 정보 입력창 및 화면 테스트 수정

- 입력창(TextField) 테스트 로직 수정
  - 입력과 포커스 상태에 따른 화면 요소 표시 확인
  - `VisualTransformation` 적용 시 화면 표시 확인
  - 오류 메시지 표시 확인
  - 최대 글자 수 입력 제한 확인 (소유자 이름)
- 카드 등록 화면(CardRegistrationScreen) 테스트 로직 수정
  - 초기 화면 및 정보 입력 후 버튼 상태 확인
- 카드 목록 화면(CardsScreen) 테스트 로직 수정
  - 빈 목록, 단일/다중 카드 상태 화면 표시 확인
- `CardholderNameUiModel` 생성 시 `maxLength` 인자 추가

moondev03 committed:
refactor: CardExpirationDate를 value class로 변경

moondev03 committed:
test: `ComposeContentTestRule` 확장 함수 추가

`onNodeWithRoleAndContentDescription` 함수를 추가하여 역할(Role)과 내용 설명(ContentDescription)으로 노드를 찾을 수 있도록 함.

moondev03 committed:
feat: Modifier 확장 함수 `semanticsContentDescription` 추가

`contentDescription` 설정을 위한 `semanticsContentDescription` Modifier 확장 함수를 추가합니다. 이 함수는 String 리소스 ID를 받아 현재 `LocalContext`를 사용하여 `contentDescription`을 설정합니다.

moondev03 committed:
refactor: 카드 만료일 UI 모델 생성 로직 수정

- `CardExpirationDate`의 `yearMonth` 대신 `value` 속성을 사용하여 포맷팅하도록 변경

moondev03 committed:
feat: 접근성 향상을 위한 Content Description 추가

- 각 카드 정보 입력 TextField에 contentDescription 추가
- 카드 등록 화면 TopAppBar의 완료 아이콘 contentDescription 수정

moondev03 committed:
refactor: `PaymentCard` 컴포저블 개선 및 프리뷰 추가

moondev03 committed:
refactor: `CardholderNameUiModel`의 `cardholderNameValue`를 `name`으로 변경

- `cardholderName` 속성을 `displayedName`으로 변경하여 명확성 향상

moondev03 committed:
feat: IntRange 확장 함수 추가

- `coerceAtMost(length: Int)`: IntRange의 각 끝 값을 주어진 길이(length)를 넘지 않도록 조정합니다.

moondev03 committed:
refactor: `PaymentCardUiModel` 카드 정보 표시 방식 개선

moondev03 committed:
refactor: UiModel 필드명 변경

moondev03 committed:
test: 카드 등록 화면 테스트 코드 수정

moondev03 committed:
refactor: `UiEvent` 초기값을 `null`에서 `None`으로 변경

- `CardsScreenUiEvent`와 `CardRegistrationScreenUiEvent`에 `None` 상태 추가
- ViewModel에서 `_uiEvent` 초기값을 `None`으로 설정
- `UiEvent` 소비 후 `None`으로 재설정하여 불필요한 이벤트 처리 방지

moondev03 committed:
refactor: 카드 목록 화면 UI 개선

moondev03 committed:
refactor: 카드 목록 화면 TopAppBar의 `isRegistrationButtonEnabled` 매개변수 이름을 `showRegistrationButton`으로 변경

moondev03 committed:
refactor: `CardRegistrationScreen` 매개변수 순서 변경

moondev03 committed:
feat: Compose `runtime-livedata` 의존성 추가

moondev03 committed:
feat: LiveData.update 확장 함수 추가

moondev03 committed:
refactor: ViewModel에서 Compose 상태 관리 방식을 LiveData로 변경

moondev03 commented:
답변이 늦어서 죄송합니다 🥲

---

> 컴포저블 하나당 하나의 `UiModel`을 사용되었기에`Composable` 코드의 복잡성도 늘어난 듯한 느낌도 듭니다.

카드 정보의 글자 수 제한이나 숫자만 입력 가능 여부 같은 검증 규칙은 도메인에서 처리해야 할지, 
컴포넌트에서 처리해야 할지, 아니면 둘 다 처리해야 할지 계속 혼란스러워 여러 번 구조를 수정하게 되었습니다.

최종적으로는 현재 컴포넌트에서 별도의 입력 검증을 수행하지 않고 도메인 모델이 책임지도록 했습니다.
각 카드 정보 TextField는 원시 타입으로 값과 에러 메시지만 받도록 구성했습니다. 
이렇게 구성하면 도메인 규칙이 변경되더라도 하위 컴포넌트의 수정을 최소화할 수 있다고 생각했습니다.
(예: 입력 가능한 최대 글자 수, 숫자 여부 등)

다만, 현재 구조에서는 각 카드 정보 TextField가 라벨과 플레이스홀더, VisualTransformation 정도만 다르기 때문에, 
사실상 OutlineTextField를 바로 사용해도 무방하다고도 느껴집니다.

지오 입장에서는 라벨과 플레이스홀더 정도만 다르더라도 별도의 컴포넌트로 분리하는 것이 좋다고 생각하시나요?

<br>

> ViewModel은 Activity에서 들고 있어야 하는가, Screen(Composable)에서 들고 있어야 하는가?

저는 ViewModel의 위치가 생명주기 범위를 어떻게 가져가느냐에 따라 달라진다고 생각합니다.
- 여러 Screen에서 공통으로 사용해야 하는 경우라면 Activity에서 관리하는 것이 적절하고
- 특정 Screen에서만 쓰이고 공유할 필요가 없다면 Screen 내부에서 생성하는 편이 맞다고 보았습니다.

현재는 Screen의 매개변수로 ViewModel을 전달하고 있는데, 이렇게 하면 추후 Screen이 다른 Activity에서 재사용될 때도 외부에서 ViewModel을 주입해 공유할 수 있는 유연성을 확보할 수 있다고 생각했습니다.

<br>

> ViewModel에서 androidx.compose.runtime에 대한 의존성을 갖고 있어도 될까?

현재 ViewModel에서 UiState를 State로 관리하고 있기 때문에 androidx.compose.runtime에 대한 의존성을 가지고 있습니다.
다만 저 역시 ViewModel이 Compose에 의존성을 직접 갖는 것이 어색하다고 생각합니다.

일반적으로 ViewModel은 Compose와 분리된 레이어로 두는 편이 권장되기 때문에, 
State 대신 LiveData로 관리하고 Composable 쪽에서 observeAsState()로 변환하는 방식으로 수정했습니다.

[58bd008](https://github.com/woowacourse/android-payments/pull/34/commits/58bd008db538cd6104d8e23ada3fd14700652a2a)

moondev03 committed:
refactor: `rememberCardAddLauncher` 에서 `viewModel` 대신 `onCardAdded` 람다 사용

giovannijunseokim commented:
> 지오 입장에서는 라벨과 플레이스홀더 정도만 다르더라도 별도의 컴포넌트로 분리하는 것이 좋다고 생각하시나요?

네. 라벨과 플레이스홀더 등을 보여주는 것을 해당 컴포넌트의 책임으로 보고, 해당 책임을 테스트하기 위함입니다. 😁 

### 분리하는 이유

- **함수의 크기가 작아진다.**
- 각 컴포넌트에서 사용하는 **상수를 컴포넌트 내부로** 감출 수 있다.
- 다른 컴포넌트에 영향받지 않고 **독립적으로 테스트**할 수 있다.

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/CardExpirationDateTextFieldTest.kt:
```diff
@@ -7,98 +7,132 @@ import androidx.compose.runtime.setValue
-        composeTestRule
-            .onNodeWithTag(TEST_TAG)
-            .performTextInput("123")
+    fun `에러메시지가_있으면_오류_텍스트가_보인다`() {
```
에러 메시지와 오류 텍스트는 어떤 차이인가요?

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/PaymentCardTest.kt:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    @Test
+    fun `카드번호는_앞의_8자리만_표시되고_이후는_마스킹된다`() {
+        composeTestRule
+            .onNodeWithText("1234 - 5678 - **** - ****")
+            .assertIsDisplayed()
+    }
```
`VisualTransform`같은 경우는 인자로 제공할 수 있고 전달된 `VisualTransform`이 잘 적용됨을 테스트했다면,
여기에선 항상 8자리 이후가 마스킹되도록 구현되었습니다.

둘은 어떤 차이가 있다고 생각하셔서 다르게 구현된걸까요? 👀

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenTest.kt:
```diff
@@ -0,0 +1,92 @@
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    @Test
+    fun `초기_화면에서_카드_등록_버튼은_비활성화된다`() {
+        // given
+        setup()
+        val cardRegistrationButton =
+            composeTestRule.onNodeWithRoleAndContentDescription(Role.Button, "카드 등록")
+
+        // then
+        assertAll(
+            { cardRegistrationButton.assertIsDisplayed() },
+            { cardRegistrationButton.assertIsNotEnabled() },
```
버튼이 비활성화되는 것만 검증하는 것이 아니라, 시각적으로 표시되는지도 검증하고 있는 것 같아요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/CardExpirationDate.kt:
```diff
@@ -0,0 +1,39 @@
+) {
+    fun isExpired(now: YearMonth = YearMonth.now()): Boolean = value < now
+
+    sealed class CardExpirationDateError(
```
sealed class로 만들어주신 이유가 무엇인가요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/CardExpirationDate.kt:
```diff
@@ -0,0 +1,39 @@
+    sealed class CardExpirationDateError(
+        message: String,
+    ) : Exception(message) {
+        class ExceedsLength : CardExpirationDateError("만료일은 ${VALID_CARD_EXPIRATION_DATE_LENGTH}자리여야 합니다. (입력값이 너무 깁니다)")
+
+        class InsufficientLength : CardExpirationDateError("만료일은 ${VALID_CARD_EXPIRATION_DATE_LENGTH}자리여야 합니다. (입력값이 너무 짧습니다)")
+
+        class NonDigit : CardExpirationDateError("만료일은 숫자만 입력할 수 있습니다.")
+
+        class UnsupportedDate : CardExpirationDateError("존재하지 않거나 유효하지 않은 만료일입니다.")
```
`object`로 선언할 수도 있을 것 같은데, 클래스로 선언한 이유가 궁금해요. 👀

giovannijunseokim replied:
또한 대부분의 `Exception`의 서브 클래스들은 `Exception`을 postfix로 붙이는 것 같은데, 여기에도 반영해볼 수 있을까요?

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/CardNumber.kt:
```diff
@@ -0,0 +1,30 @@
+    sealed class CardNumberError(
+        message: String,
+    ) : Exception(message) {
+        class ExceedsLength : CardNumberError("카드 번호는 정확히 ${VALID_CARD_NUMBER_LENGTH}자리여야 합니다. (입력값이 너무 깁니다)")
+
+        class InsufficientLength : CardNumberError("카드 번호는 정확히 ${VALID_CARD_NUMBER_LENGTH}자리여야 합니다. (입력값이 너무 짧습니다)")
```
카드 번호의 자리수가 일치하지 않는 `Exception`으로 하나의 `Exception`을 만들 수 있을 것 같은데, 더욱 세부적으로 나눠 얻을 수 있는 이점이 무엇일까요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -1,67 +1,53 @@
+import woowacourse.payments.ui.extension.semanticsContentDescription
 
 @Composable
 fun CardExpirationDateTextField(
     cardExpirationDate: String,
     onCardExpirationDateChanged: (String) -> Unit,
+    errorMessage: String?,
     modifier: Modifier = Modifier,
-    errorMessage: String? = null,
+    visualTransformation: VisualTransformation = rememberDefaultCardExpirationDateVisualTransformation(),
 ) {
```
어떤 에러 메시지를 띄울지는 해당 컴포넌트 내부에서 결정할 수도 있을 것 같아요. `VisualTransformation`도 마찬가지입니다. 외부로부터 주입받도록 하신 이유가 궁금해요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardholderNameTextField.kt:
```diff
@@ -1,49 +1,57 @@
-            if (CARDHOLDER_NAME_VALIDATION_REGEX.matches(newName).not()) return@OutlinedTextField
-
-            onCardholderNameChanged(newName.uppercase())
+            if (newValue.length <= maxLength) onCardholderNameChanged(newValue)
```
`maxLength`에 관한 내용은 `onCardholderNameChanged`에 포함시키도록 구현할 수도 있을 것 같습니다.

외부에서 준 `onCardholderNameChanged`가 그대로 적용되지 않는다면 이 컴포넌트를 사용하는 개발자 입장에서 혼란을 느낄 수 있지 않을까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,39 +1,143 @@
 
 @Composable
-fun PaymentCard(modifier: Modifier = Modifier) {
+fun PaymentCard(
+    paymentCardUiModel: PaymentCardUiModel,
+    modifier: Modifier = Modifier,
+    backgroundColor: Color = Color.DarkGray,
+    cornerRadius: Int = 5,
+) {
```
https://github.com/woowacourse/android-payments/pull/34#issuecomment-3281890273
> 각 카드 정보 TextField는 원시 타입으로 값과 에러 메시지만 받도록 구성했습니다.

`PaymentCard`는 어떤 점에서 `TextField`와 다르기에 `UiModel`이 사용된 것인지 궁금합니다. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,39 +1,143 @@
+    }
+}
+
+@Composable
+private fun PaymentCardText(
+    text: String,
+    modifier: Modifier = Modifier,
+) {
+    Text(
+        text = text,
+        color = Color.White,
+        fontSize = 12.sp,
+        fontWeight = FontWeight.W500,
+        letterSpacing = 1.sp,
+        lineHeight = 1.em,
+        modifier = modifier,
+    )
+}
```
`PaymentCard` 내부의 카드 번호만 굵기 처리가 변경된다면 어떻게 될까요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,39 +1,143 @@
+    )
+}
+
+@Composable
+private fun PaymentCardInfoBlock(
```
`Block`이라는 postfix는 어떤 기준으로 붙인 것인지 궁금합니다! 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/extension/IntRangeExt.kt:
```diff
@@ -0,0 +1,3 @@
+package woowacourse.payments.ui.extension
+
+fun IntRange.coerceAtMost(length: Int): IntRange = (first.coerceAtMost(length)..last.coerceAtMost(length - 1))
```
https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.ranges/coerce-at-most.html
> Ensures that this value is not greater than the specified [maximumValue](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.ranges/coerce-at-most.html).

Kotlin 공식문서에서 설명하는 `coerceAtMost`와 다른 역할을 하고 있는 것 같아요. 정수 범위는 `length`와 비교될 수 없다고 생각되는데, 어떻게 생각하시나요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/extension/LiveDataExt.kt:
```diff
@@ -0,0 +1,7 @@
+
+import androidx.lifecycle.MutableLiveData
+
+fun <T : Any> MutableLiveData<T>.update(transform: T.() -> T) {
+    value?.let { safeValue -> value = safeValue.transform() }
+}
```
해당 함수를 통해서 다음과 같이 작성할 수 있겠네요. 👍 

```kotlin
// AS-IS
_uiState.value = uiState.value?.copy(cardExpirationDate = newExpirationDate)

// TO-BE
_uiState.update { copy(cardExpirationDate = newExpirationDate) }
```

하지만 이렇게 되면 컨벤션 등을 통해 `setValue` 대신 `update` 함수의 사용을 강제해야 할 것 같아요. 
그렇지 않으면 프로젝트의 전반적인 일관성이 무너지니까요. 🤔 

해당 함수 사용을 강제하는 컨벤션이 필요한 정도로 해당 함수가 꼭 필요할까요? 
해당 함수를 사용할 때의 장점은 무엇인가요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/extension/ModifierExt.kt:
```diff
@@ -0,0 +1,18 @@
+import androidx.compose.ui.semantics.contentDescription
+import androidx.compose.ui.semantics.semantics
+
+@Composable
+fun Modifier.semanticsContentDescription(
+    @StringRes contentDescriptionResId: Int,
+): Modifier {
+    val context = LocalContext.current
+    return semantics {
+        contentDescription = context.getString(contentDescriptionResId)
+    }
+}
```
이 함수 또한 `MutableLiveData.update`와 유사한 것 같아요.

해당 함수를 사용하는 것이 컨벤션을 세울 정도의 장점이 있는지 궁금합니다. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/model/CardExpirationDateUiModel.kt:
```diff
@@ -0,0 +1,33 @@
+import woowacourse.payments.domain.CardExpirationDate
+import java.time.format.DateTimeFormatter
+
+@Parcelize
+data class CardExpirationDateUiModel(
+    val expirationDate: String = "",
+    val state: State = State.NOT_FILLED,
+) : Parcelable {
+    @IgnoredOnParcel
+    val isValid: Boolean = state == State.VALID
```
이미 `state: State`가 `public`하게 노출되고 있습니다.
`isValid`가 필요한 프로퍼티일까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,188 @@
+@Composable
+fun CardsScreen(viewModel: CardsScreenViewModel = rememberCardsScreenViewModel()) {
+    val context = LocalContext.current
+    val uiState by viewModel.uiState.observeAsState(CardsUiState.EMPTY)
+    val uiEvent by viewModel.uiEvent.observeAsState(CardsScreenUiEvent.None)
```
초기값은 이곳과 ViewModel 중 어디에서 결정되는 것이 좋을까요? 🤔 

현재 `uiState`의 경우 `ViewModel`과 `Screen` 모두에서 초기값을 설정해주는 것 같고,
`uiEvent`는 `Screen`에서 설정해주는 것 같아요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationActivity.kt:
```diff
@@ -0,0 +1,33 @@
+        setContent {
+            AndroidpaymentsTheme {
+                CardRegistrationScreen(
+                    onBackClick = { finish() },
```
```suggestion
                    onBackClick = ::finish,
```

위와 같이 고쳐볼 수도 있겠네요. 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreen.kt:
```diff
@@ -8,49 +8,49 @@ import androidx.compose.foundation.layout.padding
+fun CardRegistrationScreen(
+    viewModel: CardRegistrationScreenViewModel = rememberCardRegistrationScreenViewModel(),
+    onRegistrationComplete: (PaymentCardUiModel) -> Unit,
+    onBackClick: () -> Unit,
```
`onBackClick`, `onRegistrationComplete`을 내부에서 결정할 수도 있었을 것 같아요. (`LocalActivity`)

인자로 열어두고 `Activity`에서 구현하신 이유가 궁금해요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreen.kt:
```diff
@@ -8,49 +8,49 @@ import androidx.compose.foundation.layout.padding
-                snackbarState.showSnackbar(uiEvent.message.asString(context))
+            is CardRegistrationScreenUiEvent.None -> Unit
+            is CardRegistrationScreenUiEvent.RegisteredCard -> {
+                (uiEvent as? CardRegistrationScreenUiEvent.RegisteredCard)
+                    ?.paymentCard
+                    ?.let(onRegistrationComplete)
```
이곳에서 형변환이 실패하는 경우도 생길 수 있을까요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenUiState.kt:
```diff
@@ -3,25 +3,57 @@ package woowacourse.payments.ui.screen.registration
+
+    @IgnoredOnParcel
+    val cardNumberErrorMessageResId: Int? =
+        when (cardNumber.state) {
+            CardNumberUiModel.State.INVALID -> R.string.common_invalid_format_error_message
+            else -> null
+        }
```
`cardNumber.state`의 자료형은 `enum`인데, `enum`의 장점을 누리지 못하고 있는 것 같아요. 👀 

아니면 새로운 상태가 추가될 일이 없을것이라 판단하여 구현된 것일까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenViewModel.kt:
```diff
@@ -1,62 +1,166 @@
 ) {
-    var uiState by mutableStateOf(initialUiState)
-        private set
+    private var _uiState = MutableLiveData(initialUiState)
```
`var`이어야 할까요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenViewModel.kt:
```diff
@@ -1,62 +1,166 @@
-        private set
+    private var _uiEvent =
+        MutableLiveData<CardRegistrationScreenUiEvent>(CardRegistrationScreenUiEvent.None)
+    val uiEvent: LiveData<CardRegistrationScreenUiEvent> =
+        _uiEvent.also { _uiEvent.value = CardRegistrationScreenUiEvent.None }
```
`uiEvent`에 접근할 때마다 `CardRegistrationScreenUiEvent.None`으로 재할당 해주고 있는 것 같아요. 👀 

그렇게 해주시는 의도가 무엇일까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenViewModel.kt:
```diff
@@ -1,62 +1,166 @@
-        if (!uiState.isSaveButtonEnabled) return
-        val message =
-            StringResWithParams(R.string.card_registration_screen_registration_card_success)
-        uiEvent = CardRegistrationScreenUiEvent.ShowSnackbar(message)
+        if (_uiState.value?.isSaveButtonEnabled == false) return
+        val paymentCard = _uiState.value?.toPaymentCard() ?: return
+        _uiEvent.update { CardRegistrationScreenUiEvent.RegisteredCard(paymentCard) }
```
뷰에 상태(`isSaveButtonEnabled`)에 의존해서 동작이 변경되고 있어요.

의존하지 않고도 구현할 수 있을까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenViewModel.kt:
```diff
@@ -1,62 +1,166 @@
+        }
+    }
+
+    private fun CardRegistrationScreenUiState.toPaymentCard() =
+        PaymentCardUiModel(
+            number = cardNumber,
+            expirationDate = cardExpirationDate,
+            cardholderName = cardholderName,
+        )
```
`toPaymentCardUiModel`같은 경우 `public`하게 사용되고 있는 것에 반해, `toPaymentCard`는 `private`하게 사용되고 있어요.

그 차이가 무엇인가요? 👀

giovannijunseokim reviewed (approved):
미션 구현 고생 많으셨습니다. 🔥 

이번 미션은 여기서 마무리하려고 하다 보니 최대한 리뷰를 많이 남겨두었어요. 😁 

한번 씩 읽어주시고, 다음 미션에 선택적으로 반영해주시면 될 것 같습니다.
추가적으로 궁금한 점 있으시면 언제든 DM 또는 오프라인으로 질문해주세요! 🫡
